File: app1/src/utils/participantNaming.js
/**
 * Participant folder: NAME_PARTS_NNN (numeric suffix only).
 * Examples: DEVIKA_392, ANJANA_M_562, AMITHA_SHAJI_K_234
 */
export function numericParticipantSuffix(participantId) {
  const source = String(participantId || "0");
  let crc = 0xffffffff;
  for (let i = 0; i < source.length; i += 1) {
    crc ^= source.charCodeAt(i);
    for (let j = 0; j < 8; j += 1) {
      crc = (crc >>> 1) ^ (0xedb88320 & -(crc & 1));
    }
  }
  return String((crc ^ 0xffffffff) % 1000).padStart(3, "0");
}

export function buildParticipantFolderName({ name, participantId }) {
  const parts = String(name || "participant")
    .trim()
    .split(/\s+/)
    .map((part) => part.replace(/[^a-zA-Z0-9]/g, ""))
    .filter(Boolean);
  const nameToken = parts.map((part) => part.toUpperCase()).join("_") || "PARTICIPANT";
  return `${nameToken}_${numericParticipantSuffix(participantId)}`;
}

/**
 * Pose folder: PoseName_PoseID (e.g. Triangle_Pose_STA-05-I)
 */
export function buildPoseFolderName(poseName, poseId) {
  const sanitize = (value, fallback) => {
    const cleaned = String(value || fallback)
      .trim()
      .replace(/[^a-zA-Z0-9_-]/g, "_")
      .replace(/_+/g, "_")
      .replace(/^_|_$/g, "");
    return cleaned || fallback;
  };
  const namePart = sanitize(poseName, "Pose").replace(/_/g, "_");
  const idPart = sanitize(poseId, "POSE");
  return `${namePart}_${idPart}`;
}

File: backend/participant_naming.py
"""Participant folder naming: NAME_PARTS_NNN (numeric suffix only)."""

from __future__ import annotations

import re
import zlib


def numeric_participant_suffix(participant_id: str | None) -> str:
    """Deterministic 3-digit numeric suffix from participant id (digits only)."""
    source = str(participant_id or "0")
    value = zlib.crc32(source.encode("utf-8")) & 0xFFFFFFFF
    return f"{value % 1000:03d}"


def participant_folder_name(
    participant_name: str | None,
    participant_id: str | None,
) -> str:
    """
    Examples:
      Devika -> DEVIKA_392
      Anjana M -> ANJANA_M_562
      Amitha Shaji K -> AMITHA_SHAJI_K_234
    """
    raw_name = (participant_name or "participant").strip()
    parts = [
        re.sub(r"[^a-zA-Z0-9]", "", token)
        for token in re.split(r"\s+", raw_name)
        if token and re.sub(r"[^a-zA-Z0-9]", "", token)
    ]
    name_token = "_".join(p.upper() for p in parts) if parts else "PARTICIPANT"
    suffix = numeric_participant_suffix(participant_id)
    return f"{name_token}_{suffix}"
